### PR TITLE
Add cmake shell and init functions

### PR DESCRIPTION
**Pull request by chuckatkins**
Similar to before, but with working tests

### CODE REVIEW

**Comment by chuckatkins** [comment]
@rtmclay I went to add some docs after fixing the tests but it seems that none of the other non-shell *shell* environments have docs either, i.e. lisp, perl, python, or R.  Should I just add a new section in the index for "integration with scripting languages"?

**Comment by chuckatkins** [comment]
@rtmclay, also, I bypassed the need for the cmake-path issue in Hermes.  It ended up being more involves because of the additional path re-writing in common_functions.sh and ended up breaking quite a few other tests.  I worked around it on the lua side.

**Comment by chuckatkins** [comment]
ping?

File: init/cmake.in [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    set(${LMOD_ARGS_OUTPUT_VARIABLE} "${err_var}" PARENT_SCOPE)
+  endif()
+  if(LMOD_ARGS_RESULT_VARIABLE)
+    set(${LMOD_ARGS_RESULT_VARIABLE} ${ret_var} PARENT_SCOPE)
+  endif()
+endfunction(module)
+
+# Retrieve the currently loaded modules
+function(module_list out_var)
+  cmake_policy(SET CMP0007 NEW)
```
Comment (wpoely86) [inline init/cmake.in]: what does this do? Maybe add a comment?

File: init/cmake.in [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  list(REMOVE_ITEM ${out_var} "No modules loaded")
+  if(${out_var})
+    list(FILTER ${out_var} EXCLUDE REGEX "^(.*:)?$")
+  endif()
+
+  set(${out_var} ${${out_var}} PARENT_SCOPE)
+endfunction()
+
+# Swap one module for another
+function(module_swap out_mod in_mod)
```
Comment (wpoely86) [inline init/cmake.in]: Add this to the documentation on Line 40-?

File: rt/cmake/cmake.tdesc [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+-- Skip the CMake test if we can't find a new enough version fo CMake
+local skip_test = true
+local cmake_path = "/bin/false"
+local cmake_ret = os.execute("cmake --version > cmake_version.txt 2>&1")
+if cmake_ret then
+   local cmake_file = io.open("cmake_version.txt")
+   local cmake_line = cmake_file:read()
+   local cmake_ver = string.match(cmake_line, "cmake version (.*)")
+   if cmake_ver >= "3.6" then
```
Comment (wpoely86) [inline rt/cmake/cmake.tdesc]: does this work reliable? Better use `parseVersion`? ( see `src/parseVersion.lua`)

File: rt/cmake/cmake.tdesc [hunk 4]
```diff
@@ -0,0 +1,102 @@
+-- -*- lua -*-
+
+-- Skip the CMake test if we can't find a new enough version fo CMake
+local skip_test = true
+local cmake_path = "/bin/false"
+local cmake_ret = os.execute("cmake --version > cmake_version.txt 2>&1")
```
Comment (wpoely86) [inline rt/cmake/cmake.tdesc]: Aren't there cleaner ways of checking that? `cmake -E capabilities` or dump a file with
```
cmake_minimum_required (VERSION 3.6.0)
```
and do `cmake -P` with return code check?

File: init/cmake.in (hunk 1 above)
Comment (chuckatkins) [inline init/cmake.in]: Done

File: init/cmake.in (hunk 2 above)
Comment (chuckatkins) [inline init/cmake.in]: Done

File: rt/cmake/cmake.tdesc (hunk 4 above)
Comment (chuckatkins) [inline rt/cmake/cmake.tdesc]: The `capabilities` command isn't sufficient.  The requirement for 3.6.0 is the usage of the FILTER sub command for list, which wasn't present until 3.6.  Whether we parse the version in lua or use a cmake script with return code, I suppose either way, something is dumping to a file and parsing for the version check.  At least using the suggestged cmake -P method, the version will be parsed in the same way by cmake as in the init script.  I'll adjust accordingly.

File: rt/cmake/cmake.tdesc (hunk 3 above)
Comment (chuckatkins) [inline rt/cmake/cmake.tdesc]: Will eliminate in leiu of cmake_minimum_required

File: rt/cmake/cmake.tdesc (hunk 4 above)
Comment (chuckatkins) [inline rt/cmake/cmake.tdesc]: Done

**Comment by chuckatkins** [comment]
Woohoo!  Thanks!

**Comment by rtmclay** [comment]
Thanks for the patch.  I know this was a lot of work.  Yes please add a new section in the index for "integration with scripting languages"

Thanks!

**Comment by chuckatkins** [comment]
The actual lmod code was pretty straight forward, most of the work was just
getting a handle on the testing infrastructure.  I'll try to get a new pr
up next week with the initial documentation for cmake and stubs for the
other languages.

----------
Chuck Atkins
Staff R&D Engineer, Scientific Computing
Kitware, Inc.
